File: 1.21.6/src/main/java/com/bleudev/nine_lifes/util/RegistryUtils.java
package com.bleudev.nine_lifes.util;

import com.mojang.serialization.MapCodec;
import net.minecraft.entity.Entity;
import net.minecraft.entity.EntityType;
import net.minecraft.entity.effect.StatusEffect;
import net.minecraft.entity.effect.StatusEffectInstance;
import net.minecraft.item.consume.ConsumeEffect;
import net.minecraft.network.RegistryByteBuf;
import net.minecraft.network.codec.PacketCodec;
import net.minecraft.potion.Potion;
import net.minecraft.registry.Registries;
import net.minecraft.registry.Registry;
import net.minecraft.registry.RegistryKey;
import net.minecraft.registry.RegistryKeys;
import net.minecraft.registry.entry.RegistryEntry;
import net.minecraft.util.Identifier;
import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.NotNull;

import static com.bleudev.nine_lifes.NineLifesConst.MOD_ID;

public class RegistryUtils {
    @Contract("_ -> new")
    public static @NotNull Identifier getId(@NotNull String name) {
        return Identifier.of(MOD_ID, name);
    }

    @Contract("_, _ -> new")
    public static @NotNull RegistryEntry<StatusEffect> registerStatusEffect(String name, @NotNull StatusEffect effect) {
        return RegistryEntry.of(Registry.register(Registries.STATUS_EFFECT, getId(name), effect));
    }

    public static <T extends ConsumeEffect> ConsumeEffect.Type<T> registerConsumeEffectType(String name, MapCodec<T> codec, PacketCodec<RegistryByteBuf, T> packetCodec) {
        return Registry.register(Registries.CONSUME_EFFECT_TYPE, getId(name), new ConsumeEffect.Type<>(codec, packetCodec));
    }

    public static RegistryEntry<Potion> registerPotion(String name, StatusEffectInstance... effects) {
        return Registry.registerReference(Registries.POTION, getId(name), new Potion(name, effects));
    }

    public static <T extends Entity> EntityType<T> registerEntity(String name, @NotNull EntityType.Builder<T> type) {
        return Registry.register(
            Registries.ENTITY_TYPE,
            getId(name),
            type.build(RegistryKey.of(RegistryKeys.ENTITY_TYPE,
                                      getId(name)))
        );
    }
}

File: 26.1.2/src/main/kotlin/com/bleudev/nine_lifes/util/RegistryUtils.kt
package com.bleudev.nine_lifes.util

import com.bleudev.nine_lifes.MOD_ID
import com.mojang.serialization.MapCodec
import net.minecraft.core.Holder
import net.minecraft.core.Registry
import net.minecraft.core.registries.BuiltInRegistries
import net.minecraft.core.registries.Registries
import net.minecraft.network.RegistryFriendlyByteBuf
import net.minecraft.network.codec.StreamCodec
import net.minecraft.resources.Identifier
import net.minecraft.resources.ResourceKey
import net.minecraft.world.effect.MobEffect
import net.minecraft.world.entity.Entity
import net.minecraft.world.entity.EntityType
import net.minecraft.world.item.Item
import net.minecraft.world.item.consume_effects.ConsumeEffect

fun createIdentifier(path: String): Identifier = Identifier.fromNamespaceAndPath(MOD_ID, path)

private fun <T : Any> ResourceKey<Registry<T>>.key(name: String) =
    ResourceKey.create(this, createIdentifier(name))
private fun <V : Any, T : V> Registry<V>.register(id: Identifier, element: T) =
    Registry.register(this, id, element)
private fun <V : Any, T : V> Registry<V>.register(name: String, element: T) =
    register(createIdentifier(name), element)

fun registerMobEffect(name: String, effect: MobEffect): Holder<MobEffect> =
    Registry.registerForHolder(BuiltInRegistries.MOB_EFFECT, createIdentifier(name), effect)

fun <T : ConsumeEffect> registerConsumeEffectType(name: String, mCodec: MapCodec<T>, sCodec: StreamCodec<RegistryFriendlyByteBuf, T>): ConsumeEffect.Type<T> =
    BuiltInRegistries.CONSUME_EFFECT_TYPE.register(name, ConsumeEffect.Type(mCodec, sCodec))

fun <T : Entity> registerEntity(path: String, type: EntityType.Builder<T>): EntityType<T> =
    BuiltInRegistries.ENTITY_TYPE.register(Identifier.parse(path) /* For compatibility */, type.build(Registries.ENTITY_TYPE.key(path)))

fun <T : Item> registerItem(name: String, itemFactory: (Item.Properties) -> T, properties: Item.Properties = Item.Properties()): T =
    BuiltInRegistries.ITEM.register(name, itemFactory(properties.setId(Registries.ITEM.key(name))))
